lib/new_relic/agent/instrumentation/aws_sdk_lambda/instrumentation.rb

```diff
@@ -0,0 +1,112 @@
+# This file is distributed under New Relic's license terms.
+# See https://github.com/newrelic/newrelic-ruby-agent/blob/main/LICENSE for complete details.
+# frozen_string_literal: true
+
+require 'json'
+
+module NewRelic::Agent::Instrumentation
+  module AwsSdkLambda
+    INSTRUMENTATION_NAME = 'aws_sdk_lambda'
+    AWS_SERVICE = 'lambda'
+    CLOUD_PLATFORM = 'aws_lambda'
+    WRAPPED_RESPONSE = Struct.new(:status_code, :has_status_code?)
+
+    def invoke_with_new_relic(*args)
+      with_tracing(:invoke, *args) { yield }
+    end
+
+    def invoke_async_with_new_relic(*args)
+      with_tracing(:invoke_async, *args) { yield }
+    end
+
+    def invoke_with_response_stream_with_new_relic(*args)
+      with_tracing(:invoke_with_response_stream, *args) { yield }
+    end
+
+    private
+
+    def with_tracing(action, *args)
+      segment = generate_segment(action, *args)
+
+      # prevent additional instrumentation for things like Net::HTTP from
+      # creating any segments that may appear as redundant / confusing
+      NewRelic::Agent.disable_all_tracing do
+        begin
+          response = yield
+          process_response(response, segment)
+          response
+        rescue => e
+          NewRelic::Agent.notice_error(e)
+          raise
```

Comment:
Is this intended to capture errors both raised by the agent and raised by the customer's invocation of the AWS lambda gem?

Reply:
Ah interesting. I'm a little concerned about stopping the customer's application if the agent raises an error. 

Would a `NewRelic::Agent::Tracer.capture_segment_error` block work around the yield to catch customer-raised errors? Totally understand if not, this seems like a pretty unique scenario compared to our other instrumentation.

Reply:
Great call. I was thinking of it only from the perspective of both sources of exceptions sharing a `rescue`. They certainly should not share the `raise` which was intended only for client sdk raised exceptions. Will refactor!

Reply:
Refactored with [0a1527d](https://github.com/newrelic/newrelic-ruby-agent/pull/2926/commits/0a1527d8012cfb34c3950d34180d0974b51ab9ed) (with co-author!)